src/ngraph/pass/liveness.cpp

```diff
@@ -102,79 +101,31 @@ bool pass::Liveness::run_on_function(shared_ptr<ngraph::Function> function)
 
         for (descriptor::Tensor* tensor_decl : all_tensor_decls)
         {
-            if (!contains(currently_live, tensor_decl))
+            if (currently_live.find(tensor_decl) == currently_live.end())
             {
                 // this is the last node that value is seen in
                 // delete it at the end of the op
                 currently_live.insert(tensor_decl);
-                free_tensor_decls.insert(tensor_decl);
+                if (output_tensors.find(tensor_decl) == output_tensors.end())
+                {
+                    // Don't free output tensors
+                    free_tensor_decls.insert(tensor_decl);
+                }
             }
         }
 
-        node->liveness_live_list = currently_live;
         for (descriptor::Tensor* output_decl : output_tensor_decls)
         {
-            if (contains(currently_live, output_decl))
+            auto currently_live_it = currently_live.find(output_decl);
+            if (currently_live_it != currently_live.end())
             {
                 new_tensor_decls.insert(output_decl);
-                currently_live.erase(output_decl);
+                currently_live.erase(currently_live_it);
             }
         }
         node->liveness_free_list = free_tensor_decls;
         node->liveness_new_list = new_tensor_decls;
     }
 
-    // Anything marked as output must remain live for the remainder of the graph
-    // Add outputs to live_list and remove from free_list
-    unordered_set<descriptor::Tensor*> outputs;
-    unordered_set<descriptor::Tensor*> seen;
-    for (shared_ptr<Node> node : ops)
-    {
-        for (descriptor::Tensor* tensor : node->liveness_live_list)
-        {
-            if (contains(output_tensors, tensor))
-            {
-                outputs.insert(tensor);
-            }
-        }
-        for (descriptor::Tensor* tensor : outputs)
-        {
-            node->liveness_live_list.insert(tensor);
-            node->liveness_free_list.erase(tensor);
-
-            if (contains(node->liveness_new_list, tensor))
-            {
-                if (contains(seen, tensor))
-                {
-                    node->liveness_new_list.erase(tensor);
-                }
-                else
-                {
-                    seen.insert(tensor);
-                }
-            }
-        }
-    }
-
-    // validate_liveness(ops);
     return false;
 }
-
-void pass::Liveness::validate_liveness(const list<Node*>& ops)
-{
```

Comment:
i think we do need some sanity checks. We could just keep track of dead tensors and check that no node is using them as inputs later on?